src/RMGOutputTools.cc

```diff
@@ -0,0 +1,399 @@
+
+  for (auto hit : *hits->GetVector()) hits_map[hit->track_id].push_back(hit);
+
+  // if requested we can combine low energy tracks to reduce further file size
+  if (cluster_pars.combine_low_energy_tracks)
+    hits_map = combine_low_energy_tracks(hits_map, cluster_pars, has_distance_to_surface);
+
+  if (cluster_pars.reassign_gamma_energy)
+    redistribute_gamma_energy(hits_map, cluster_pars, has_distance_to_surface);
+
+  // create a vector of clusters of hits
+  std::vector<std::vector<RMGDetectorHit*>> hits_vector;
+
+  // keep track of the current cluster
+  // loop over trackid and then hits in each track
+  for (const auto& [trackid, input_hits] : hits_map) {
+    RMGDetectorHit* cluster_first_hit = nullptr;
+
+    for (auto hit : input_hits) {
+
+      if (!hit) continue;
+
+      // within track clustering
+
+      // if the hit is:
+      // - the first in a track
+      // - in a new detector (compared to the current cluster)
+      // - more than the time-threshold since the first hit of the cluster
+      // then we need to start a new cluster.
+
+      bool start_new_cluster = (cluster_first_hit == nullptr) or
+                               (hit->track_id != cluster_first_hit->track_id) or
+                               (hit->detector_uid != cluster_first_hit->detector_uid) or
+                               (std::abs(hit->global_time - cluster_first_hit->global_time) >
+                                   cluster_pars.cluster_time_threshold);
+      // check distances and if the track moved from surface to bulk
+      if (!start_new_cluster) {
+        bool is_surface = has_distance_to_surface and
+                          (hit->distance_to_surface_average < cluster_pars.surface_thickness);
+        bool is_surface_first_hit =
+            has_distance_to_surface and
+            (cluster_first_hit->distance_to_surface_average < cluster_pars.surface_thickness);
+
+        // start a new cluster if the previous step was in the surface and the new is in the bulk
+        bool surface_transition = (is_surface != is_surface_first_hit);
+
+        // get the right distance to pre-cluster
+        double threshold =
+            is_surface ? cluster_pars.cluster_distance_surface : cluster_pars.cluster_distance;
+
+
+        // start a new cluster also if the distance is above the threshold
+        start_new_cluster =
+            surface_transition ||
+            (hit->global_position_average - cluster_first_hit->global_position_average).mag() >=
+                threshold;
+      }
+
+      // add the hit to the correct vector
+      if (start_new_cluster) {
+        hits_vector.push_back(std::vector<RMGDetectorHit*>());
+        hits_vector.back().push_back(hit);
+
+        cluster_first_hit = hit;
+      } else {
+        hits_vector.back().push_back(hit);
+      }
+    }
+  }
+
+  // create a container for the output hits
+  auto out = std::shared_ptr<RMGDetectorHitsCollection>();
```

Comment:
```suggestion
  auto out = std::make_shared<RMGDetectorHitsCollection>();
```

with the other you constructed a nullptr, I guess?